### DISCUSSION TITLE
How to update the value in AsyncNotifier's state while keeping the loading or error state

### DISCUSSION THREAD

**Discussion by taiju59** [rrousselGit/riverpod#3891] [Q&A]
Is there a way to update the value in an AsyncNotifier's state while keeping the current loading or error state?
If possible, I’d appreciate any specific methods or alternative approaches to achieve the use case described at the end of this post.

Below is an example implementation (replaceValue) that I’d like to achieve:

```dart
state = AsyncValue.data(oldValue);
state = AsyncValue.error(Exception(), StackTrace.current);

// At this point:
// state is AsyncError
// state.hasError == true
// state.valueOrNull == oldValue

replaceValue((currentValue) => newValue);

// Expected state:
// state is AsyncError
// state.hasError == true
// state.valueOrNull == newValue
```

Similar implementation:

```dart
state = AsyncValue.data(newValue);

// :x: state is! AsyncError
// :x: state.hasError != true
```

Use case:
The scenario I’m trying to address is as follows:

1. A Notifier holds product information, including a "favorite" status.
2. This Notifier listens to favorite registration events triggered from other screens or devices (e.g., web).
3. When a favorite registration event occurs, the "favorite" status of the product information needs to be updated.

<details><summary>Code Example</summary>

```dart
@freezed
class Product with _$Product {
  const factory Product({
    required String name,
    required bool favorite,
  }) = _Product;
}

@riverpod
class ProductNotifier extends _$ProductNotifier {
  @override
  Future<Product> build(String name) async {
    ref.listen(favoriteEventProvider, (_, event) {
      _replaceFavorite(event.name, event.favorite);
    });

    // In practice, fetch data from the server
    return Product(name: name, favorite: false);
  }

  void _replaceFavorite(String name, bool favorite) => replaceValue(
        (currentvalue) =>
            currentvalue.name == name ? e.copyWith(favorite: favorite) : e,
      );
}
```

</p>
</details>

In this case, I need a way to update only the "favorite" status while retaining the loading or error state of the current state.

I would greatly appreciate any advice or ideas on how to handle this.
Thank you in advance!

**Comment by snapsl** [comment]
Using [`copyWithPrevious`](https://pub.dev/documentation/riverpod/latest/riverpod/AsyncValue/copyWithPrevious.html) like this should work.

```dart
state = state.copyWithPrevious(AsyncData(newValue));
```

**Reply by taiju59** [reply]
@snapsl Thank you for your suggestion. However, this approach does not work.
`value` is not updated.

```dart
final newState = state.copyWithPrevious(AsyncData(newValue));
print(newState.valueOrNull); // This prints newValue

state = newState;
print(state.valueOrNull); // This prints oldValue
```

It seems that this happens because the AsyncNotifier's state setter internally uses copyWithPrevious.

Here’s the process as I understand it:

1. The setter calls onError. [Source code](https://github.com/rrousselGit/riverpod/blob/riverpod-v2.6.1/packages/riverpod/lib/src/async_notifier/base.dart#L210-L217)
2. onError calls asyncTransition. [Source code](https://github.com/rrousselGit/riverpod/blob/riverpod-v2.6.1/packages/riverpod/lib/src/async_notifier/base.dart#L240-L242)
3. asyncTransition calls copyWithPrevious. [Source code](https://github.com/rrousselGit/riverpod/blob/riverpod-v2.6.1/packages/riverpod/lib/src/common.dart#L29)
4. In AsyncError.copyWithPrevious, the previous value is prioritized, so the value is not changed. [Source code](https://github.com/rrousselGit/riverpod/blob/riverpod-v2.6.1/packages/riverpod/lib/src/common.dart#L536)

To illustrate point 4 with pseudocode:

```
AsyncError(newValue).copyWithPrevious(AsyncError(oldValue));
-> AsyncError(oldValue);
```

Because of this, the state does not update to the newValue as expected.

**Reply by snapsl** [reply]
Does setting the isRefresh flag help? Unfortunately, I cannot test this at the moment.

**Reply by taiju59** [reply]
No. It will not work to set isRefresh flag  either true or false.

**Comment by rrousselGit** [comment]
What you're asking is not a thing.

**Reply by taiju59** [reply]
@rrousselGit Thank you for your response. Could you clarify if what I'm asking is not supported by Riverpod, or if there’s a recommended alternative approach to achieve this?

If my question is unclear, please let me know, and I’ll provide more details about the use case or my intent.
